sql/hive/src/main/scala/org/apache/spark/sql/hive/HiveExternalCatalog.scala

```diff
@@ -616,6 +617,44 @@ private[spark] class HiveExternalCatalog(conf: SparkConf, hadoopConf: Configurat
     client.getPartition(db, table, spec)
   }
 
+  override def listPartitionsByFilter(
+      db: String,
+      table: String,
+      predicates: Seq[Expression]): Seq[CatalogTablePartition] = withClient {
+    val catalogTable = client.getTable(db, table)
+    val partitionColumnNames = catalogTable.partitionColumnNames.toSet
+    val nonPartitionPruningPredicates = predicates.filterNot {
+      _.references.map(_.name).toSet.subsetOf(partitionColumnNames)
+    }
+
+    if (nonPartitionPruningPredicates.nonEmpty) {
+        sys.error("Expected only partition pruning predicates: " +
+          predicates.reduceLeft(And))
+    }
+
+    val partitionSchema = catalogTable.partitionSchema
+
+    if (predicates.nonEmpty) {
+      val clientPrunedPartitions =
+        client.getPartitionsByFilter(catalogTable, predicates)
+      val boundPredicate =
+        InterpretedPredicate.create(predicates.reduce(And).transform {
+          case att: AttributeReference =>
+            val index = partitionSchema.indexWhere(_.name == att.name)
```

Comment:
@mallman I think there is an issue in the modified test, since the `ParquetDataWithKey` class still has single-p as a field. That would cause false positive failures.

I tried locally to reproduce a mixed-case issue, with no success, e.g.
```
sql("create external table test_mixed_case (a int) PARTITIONED BY (someString int) STORED AS parquet LOCATION '/tmp/test_mixed_case'")
spark.sqlContext.range(10).selectExpr("id as a", "id as someString").write.partitionBy("someString").mode("overwrite").parquet("/tmp/test_mixed_case")
sql("msck repair table test_mixed_case")
sql("select * from test_mixed_case where someString = 3").count()
```

seems to work fine.